src/main.rs

```diff
@@ -0,0 +1,449 @@
+        pallet_timestamp::Call::set {
+            now: u64::from(block) * SLOT_DURATION,
+        },
+    )))
+    .unwrap()
+    .unwrap();
+
+    // 3.  Set up parachain validation data
+    let parachain_validation_data = {
+        let parent_head = HeadData(prev_header.unwrap_or(parent_header).encode());
+
+        let sproof_builder = RelayStateSproofBuilder {
+            para_id: ParachainInfo::get(),
+            included_para_head: Some(parent_head.clone()),
+            current_slot: cumulus_primitives_core::relay_chain::Slot::from(2 * u64::from(block)),
+            ..Default::default()
+        };
+
+        let (relay_storage_root, proof) = sproof_builder.into_state_root_and_proof();
+
+        cumulus_pallet_parachain_system::Call::set_validation_data {
+            data: cumulus_primitives_parachain_inherent::ParachainInherentData {
+                validation_data: cumulus_primitives_core::PersistedValidationData {
+                    parent_head: Default::default(),
+                    relay_parent_number: block,
+                    relay_parent_storage_root: relay_storage_root,
+                    max_pov_size: Default::default(),
+                },
+                relay_chain_state: proof,
+                downward_messages: Default::default(),
+                horizontal_messages: Default::default(),
+            },
+        }
+    };
+
+    Executive::apply_extrinsic(UncheckedExtrinsic::new_bare(RuntimeCall::ParachainSystem(
+        parachain_validation_data,
+    )))
+    .unwrap()
+    .unwrap();
+}
+
+fn finalize_block(elapsed: Duration) -> RelayHeader {
+    #[cfg(not(feature = "fuzzing"))]
+    println!("\n  time spent: {elapsed:?}");
+
+    assert!(elapsed.as_secs() <= 2, "block execution took too much time");
+
+    #[cfg(not(feature = "fuzzing"))]
+    println!("finalizing block");
+    Executive::finalize_block()
+}
+
+fn check_invariants(_block: u32, _initial_total_issuance: Balance) {
+    for (account, info) in Account::<Runtime>::iter() {
+        let consumers = info.consumers;
+        let providers = info.providers;
+        assert!(!(consumers > 0 && providers == 0), "Invalid c/p state");
+        let max_lock: Balance = Balances::locks(&account)
+            .iter()
+            .map(|l| l.amount)
+            .max()
+            .unwrap_or_default();
+        assert_eq!(
+            max_lock, info.data.frozen,
+            "Max lock should be equal to frozen balance"
+        );
+        let sum_holds: Balance = Holds::<Runtime>::get(&account)
+            .iter()
+            .map(|l| l.amount)
+            .sum();
+        assert!(
+            sum_holds <= info.data.reserved,
+            "Sum of all holds ({sum_holds}) should be less than or equal to reserved balance {}",
+            info.data.reserved
+        );
+    }
+
+    AllPalletsWithSystem::integrity_test();
+    check_dapp_staking_invariants();
+}
+
+fn check_dapp_staking_invariants() {
```

Comment:
You should reuse what Igor wrote for try-state checks.

There are lots of various dApp staking checks we already use for on-chain data verification.